### PR TITLE
StateProofs: Relax MSS keys ephemerality

### PR DESCRIPTION
**Pull request by Aharonee**
#### Verifier

[crypto/merklesignature/merkleSignatureScheme.go:77](crypto/merklesignature/merkleSignatureScheme.go:77)

- [X]  Add `KeyLifetime` field
- [X]  [crypto/merklesignature/merkleSignatureScheme.go:253](crypto/merklesignature/merkleSignatureScheme.go:253) round down the `round` to the closest `KeyLifetime` (256) divisor
- [X]  Change the `Verifier` to contain KeyLifetime non-exported field instead of `Keylifetime()` getter? What happens to `AccountData` which is stored in DB currently? migration?
- [ ]  Implement `Copy()` method for deep-copy

#### Signer

- [X]  ~~Add `KeyLifetime` field~~
- [X]  [crypto/merklesignature/merkleSignatureScheme.go:192](crypto/merklesignature/merkleSignatureScheme.go:192) change the algorithm cfor finding merkle tree index (round down to 256 divisor)
- [ ]  ~~[crypto/merklesignature/posdivs.go:28](crypto/merklesignature/posdivs.go:28) validate round is in validity range~~
- [x]  [crypto/merklesignature/merkleSignatureScheme.go:208](crypto/merklesignature/merkleSignatureScheme.go:208) `GetKey`: should we get key valid for a specifc round here instead? ***Only used in tests it seems, should we get rid of this?***

Q: *Where does `Signer` gets its key from? Check if it is the correct one.*

A: participationDB

#### ParticipationRegistry

- [X]  [data/account/participationRegistry.go:751](data/account/participationRegistry.go:751) `GetStateProofForRound`: fetch the correct round from DB
- [ ]  ~~Update `GetForRound`~~ *no stateproof secrets here*
- [X] Update key deletion 

#### Key Generation

- [X]  [data/account/participation.go:224](data/account/participation.go:224) `FillDB`: use hard-coded 256 instead of deriving the value from CompactCertRounds consensus param
- [X]  Make sure `SignerContext.KeyLifetime` (previously `Interval`) is set correctly

#### CompactCert Participant

[data/basics/stateProofParticipant.go:61](data/basics/stateProofParticipant.go:61)

- [X]  Add `KeyLifetime` to the serialization, since it needs to be part of the committed participants tree
- [X]  [ledger/accountdb.go:1811](ledger/accountdb.go:1811) `accountDataToOnline`:
- [ ]  ~~[ledger/ledgercore/votersForRound.go:112](ledger/ledgercore/votersForRound.go:112) init `Verifier` with 256 here and the stored commitment~~

#### Refactoring

- [ ]  Rename relevant `Round` fields to `FirstValid` or something similar
- [x]  Replace `interval` instances with `keyLifetime` if appropriate

#### Tests

- [x]  Update all failing unit tests
- [x]  Update all failing e2e tests

### CODE REVIEW

**Review by algoidan — COMMENTED** [review]
overall looks good
have some minor comments

**Comment by algoidan** [comment]
we should update the `generateMssKat` in `crypto/merklesignature/kats_test.go` to use the default instead of the consensus version (are there any more places like this?)

File: cmd/algokey/part.go [hunk 1]
```diff
@@ -161,8 +161,9 @@ func printPartkey(partkey account.Participation) {
 	fmt.Printf("Parent address:    %s\n", partkey.Parent.String())
 	fmt.Printf("VRF public key:    %s\n", base64.StdEncoding.EncodeToString(partkey.VRF.PK[:]))
 	fmt.Printf("Voting public key: %s\n", base64.StdEncoding.EncodeToString(partkey.Voting.OneTimeSignatureVerifier[:]))
-	if partkey.StateProofSecrets != nil && !partkey.StateProofSecrets.GetVerifier().IsEmpty() {
-		fmt.Printf("State proof key:   %s\n", base64.StdEncoding.EncodeToString(partkey.StateProofSecrets.GetVerifier()[:]))
+	if partkey.StateProofSecrets != nil && !partkey.StateProofSecrets.GetVerifier().MsgIsZero() {
```
[the remark this answers is not in the document]
Comment (Aharonee) [reply cmd/algokey/part.go:164]: Maybe if it was used anywhere else in the code

File: crypto/merklesignature/posdivs_test.go [hunk 2]
```diff
@@ -73,12 +73,6 @@ func TestErrors(t *testing.T) {
 	round := uint64(0)
 	require.Equal(t, errRoundNotZero, checkMerkleSignatureSchemeParams(firstValid, round, interval))
 
-	round = interval - 1
```
[the remark this answers is not in the document]
Comment (Aharonee) [reply crypto/merklesignature/posdivs_test.go:76]: `checkMerkleSignatureSchemeParams` no longer validates that

File: crypto/merklesignature/merkleSignatureScheme.go [hunk 3]
```diff
@@ -58,7 +58,7 @@ type (
 	Signer struct {
 		SigningKey *crypto.FalconSigner
 
-		// The round for which this SigningKey is related to
+		// The requested round to be signed by this Signer. Use FirstRoundInKeyLifetime in order to calculate the actual round of the key.
```
Comment (algoidan) [inline crypto/merklesignature/merkleSignatureScheme.go:61]: I think we can make that private fields.

File: crypto/merklesignature/merkleSignatureScheme.go [hunk 4]
```diff
@@ -58,7 +58,7 @@ type (
 	Signer struct {
 		SigningKey *crypto.FalconSigner
```
Comment (algoidan) [inline crypto/merklesignature/merkleSignatureScheme.go:59]: Can we make the signer fields private?

File: crypto/merklesignature/merkleSignatureScheme_test.go [hunk 5]
```diff
@@ -225,32 +269,33 @@ func TestSigning(t *testing.T) {
...
-			a.Error(err)
-		} else {
-			sig, err = signer.GetSigner(i).SignBytes(msg)
-			a.NoError(err)
-			a.NoError(signer.GetVerifier().VerifyBytes(i, msg, sig))
-		}
+	i := uint64(50)
+	for ; i < 60; i++ { // no key for these rounds (key for round 48 was not generated)
+		_, err = signer.GetSigner(i).SignBytes(msg)
+		a.Error(err)
```
Comment (algoidan) [inline crypto/merklesignature/merkleSignatureScheme_test.go:285]: can we use errorIs?

**Review by algoidan — APPROVED** [review]
looks great!
have just some minor nits

**Pull request merged**